consensus/service/src/byzantine_ledger/worker.rs

```diff
@@ -327,35 +314,44 @@ impl<
         }
 
         // Nominate values for current slot.
-        self.nominate_pending_values();
+        if self.need_nominate {
+            self.nominate_pending_values();
+        }
 
         // Process any queues consensus messages.
         self.process_consensus_msgs();
 
         // Process SCP timeouts.
-        for outgoing_msg in self.scp.process_timeouts().into_iter() {
-            (self.send_scp_message)(outgoing_msg);
+        for msg in self.scp_node.process_timeouts().into_iter() {
+            let _ = self.issue_consensus_message(msg);
         }
 
-        // See if we're done with the current slot.
-        self.attempt_complete_cur_slot();
+        if let Some(externalized_values) = self
+            .scp_node
+            .get_externalized_values(self.current_slot_index)
+        {
+            // The current slot has reached consensus.
+            self.complete_current_slot(externalized_values);
+        }
 
         // Update metrics.
-        self.update_cur_metrics();
+        self.update_current_slot_metrics();
 
         true
     }
 
-    fn process_external_requests(&mut self) -> bool {
-        for task_msg in self.receiver.try_iter() {
+    // Reads tasks from the task queue.
+    // Returns false if the worker has been asked to stop.
+    fn receive_tasks(&mut self) -> bool {
+        for task_msg in self.tasks.try_iter() {
```

Comment:
this is `try_recv` in the crate tests. Since we now have msg batching, maybe the right logic should be "process up to N new values, M new messages, and X stop messages but don't loop here forever" -- this is connected to your comment on the channel depth from earlier